File: AtCoder/AGC034/C.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main(){
    long long N, X; cin >> N >> X;
    vector<long long> b(N), l(N), u(N);
    for(int i=0;i<N;i++) cin >> b[i] >> l[i] >> u[i];
    vector<pair<long long, int>> vp;
    for(int i=0;i<N;i++) vp.emplace_back((X-b[i])*u[i] + b[i]*l[i], i);
    sort(vp.rbegin(), vp.rend());
    long long thr = 0;
    for(int i=0;i<N;i++) thr += b[i] * l[i];
    long long L = -1, R = N*X;
    while(R-L > 1){
        long long mid = (L+R)/2;
        long long sum = 0;
        int div = mid/X;
        for(int i=0;i<div;i++) sum += vp[i].first;
        long long cur = 0;
        long long m = mid%X;
        for(int i=0;i<N;i++){
            int idx = vp[i].second;
            long long add = min(m, b[idx]) * l[idx] + max(0LL, m-b[idx]) * u[idx];
            if(i < div) cur = max(cur, sum - vp[i].first + vp[div].first + add);
            else cur = max(cur, sum + add);
        }
        if(cur >= thr) R = mid;
        else L = mid;
    }
    cout << R << endl;
}

File: AtCoder/AGC034/C.rs
use std::io::Read;

fn main() {
    let mut buf = String::new();
    std::io::stdin().read_to_string(&mut buf).ok();
    let mut it = buf.split_whitespace();
    let n = it.next().unwrap().parse::<i64>().unwrap();
    let x = it.next().unwrap().parse::<i64>().unwrap();
    let mut t = Vec::new();
    let mut aoki = 0;
    for _ in 0..n {
        let b = it.next().unwrap().parse::<i64>().unwrap();
        let l = it.next().unwrap().parse::<i64>().unwrap();
        let u = it.next().unwrap().parse::<i64>().unwrap();
        let mx = b * l + (x - b) * u;
        aoki += b * l;
        t.push((b, l, u, mx));
    }
    t.sort_by_key(|a| -a.3);
    let mut low = -1;
    let mut high = n * x;
    while high - low > 1 {
        let mid = (low + high) / 2;
        let thr = (mid / x) as usize;
        let y = mid % x;
        let mut sum = 0;
        for i in 0..thr { sum += t[i].3 }
        let mut best = 0;
        for i in 0..n as usize {
            let mut cur = sum;
            if i < thr {
                cur += t[thr].3 - t[i].3;
            }
            cur += std::cmp::min(t[i].0, y) * t[i].1 + std::cmp::max(0, y - t[i].0) * t[i].2;
            best = std::cmp::max(best, cur);
        }
        if best >= aoki {
            high = mid;
        } else {
            low = mid;
        }
    }
    println!("{}", high);
}

